lightning/src/chain/channelmonitor.rs

```diff
@@ -3303,6 +3312,60 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitorImpl<Signer> {
 			}
 		}
 
+		// Fail back HTLCs on backwards channels if they expire within `TIMEOUT_FAIL_BACK_BUFFER`
+		// blocks. If we haven't seen the preimage for an HTLC by the time the previous hop's
+		// timeout expires, we've lost that HTLC, so we might as well fail it back instead of having our
+		// counterparty force-close the channel.
+		let height = self.best_block.height();
+		macro_rules! fail_soon_to_expire_htlcs {
+			($htlcs: expr) => {{
+				for (htlc, source_opt) in $htlcs {
+					// Only check forwarded HTLCs' previous hops
+					let source = match source_opt {
+						Some(source) => source,
+						None => continue,
+					};
+					let cltv_expiry = match source {
+						HTLCSource::PreviousHopData(HTLCPreviousHopData { cltv_expiry: Some(cltv_expiry), .. }) => *cltv_expiry,
+						_ => continue,
+					};
+					if cltv_expiry <= height + TIMEOUT_FAIL_BACK_BUFFER {
```

Comment:
Makes sense. I'll try and check the other implementations to make sure before ditching the buffer, since the grace period seems to be only mentioned in an example in the spec as opposed to being explicitly required AFAICT

Reply:
> Maybe, but if they dont, there's also races across when we see blocks.

Are you saying this is in addition to the grace period? Isn't this what the grace period is for?

Reply:
Other implementations grace periods:
- CLN: [Looks like they have a grace period of 1 block](https://github.com/ElementsProject/lightning/blob/8d5be61fcb57f588908c7014de8fe31c4c5953cc/lightningd/peer_htlcs.c#L2710)
- LND: [Looks like they have a grace period denoted in time behind a config flag](https://github.com/lightningnetwork/lnd/blob/cbb4c27a444103eb27afecce47a97df2e6d738db/contractcourt/channel_arbitrator.go#L1673), though it [defaults to 0](https://github.com/lightningnetwork/lnd/blob/master/config.go#L68)
- Eclair: They don't have a grace period AFAICT? [Here](https://github.com/ACINQ/eclair/blob/3547f87f664c5e956a6d13af530a3d1cb6fc1052/eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala#L363C1-L363C1) they define expired htlcs, which is called when handling a new block [here](https://github.com/ACINQ/eclair/blob/master/eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala#L2576C3-L2576C3), which directly force closes [here](https://github.com/ACINQ/eclair/blob/3547f87f664c5e956a6d13af530a3d1cb6fc1052/eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ErrorHandlers.scala#L99)

So, I've left the buffer, but replaced it with our existing `LATENCY_GRACE_PERIOD_BLOCKS`